File: Successful Pairs of Spells and Potions/successful-pairs-of-spells-and-potions.py
class Solution(object):
    def successfulPairs(self, spells, potions, success):
        """
        :type spells: List[int]
        :type potions: List[int]
        :type success: int
        :rtype: List[int]
        """
        res = []
        potions.sort()
        for spell in spells:
            min_potion = (success + spell - 1) // spell
            l, r = 0, len(potions)
            while l < r:
                middle = l + (r - l) // 2
                if potions[middle] >= min_potion:
                    r = middle
                else:
                    l = middle + 1
            res.append(len(potions) - l)
        return res

File: Successful Pairs of Spells and Potions/successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions,
                                long long success) {
        vector<int> res;
        sort(potions.begin(), potions.end());
        for (int i = 0; i < spells.size(); i++) {
            int spell = spells[i];
            long long minpotion = (success+spell-1)/spell;
            int l = 0;
            int r = potions.size(), index = 0;
            while (l < r) {
                int middle = l + (r - l) / 2;
                if (potions[middle] >= minpotion) {
                    r = middle;
                } else
                    l = middle + 1;
            }
            res.push_back(potions.size()-l);
        }
        return res;
    }
};